assets/src/stories-editor/components/right-click-menu/index.js

```diff
@@ -0,0 +1,322 @@
+/* global navigator */
+/**
+ * External dependencies
+ */
+import { castArray } from 'lodash';
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { cloneBlock, pasteHandler, serialize } from '@wordpress/blocks';
+import { useEffect, useState, useRef } from '@wordpress/element';
+import {
+	MenuGroup,
+	MenuItem,
+	NavigableMenu,
+	Popover,
+} from '@wordpress/components';
+import { withDispatch, withSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import './edit.css';
+import {
+	copyTextToClipBoard,
+	ensureAllowedBlocksOnPaste,
+} from '../../helpers';
+
+const POPOVER_PROPS = {
+	className: 'amp-story-right-click-menu__popover block-editor-block-settings-menu__popover editor-block-settings-menu__popover',
+	position: 'bottom left',
+};
+
+function useOutsideClickChecker( ref, onClose ) {
+	/**
+	 * Close the Popover if outside click was detected.
+	 *
+	 * @param {Object} event Click event.
+	 */
+	function handleClickOutside( event ) {
+		if ( ref.current && ! ref.current.contains( event.target ) ) {
+			onClose();
+		}
+	}
+
+	useEffect( () => {
+		// Handle click outside only if the the menu has been added.
+		if ( ref.current && ref.current.innerHTML ) {
+			document.addEventListener( 'mousedown', handleClickOutside );
+		}
+		return () => {
+			// Unbind when cleaning up.
+			document.removeEventListener( 'mousedown', handleClickOutside );
+		};
+	} );
+}
+
+const RightClickMenu = ( props ) => {
+	const {
+		clientIds,
+		clientX,
+		clientY,
+		copyBlock,
+		cutBlock,
+		getBlock,
+		getCopiedMarkup,
+		removeBlock,
+		duplicateBlock,
+		pasteBlock,
+	} = props;
+	const [ isOpen, setIsOpen ] = useState( true );
+
+	useEffect( () => {
+		setIsOpen( true );
+	}, [ clientIds, clientX, clientY ] );
```

Comment:
I wonder if the full `event` object should be passed here instead of these params, theoretically it's possibly that the user tries to open the menu exactly in the same point and it wouldn't open then.

Reply:
The event object will be a `SyntheticEvent`, and as those are pooled, they might match by reference, but not by contents - so using them directly as guards for an effect would probably give weird results (but most times "work").

Events have a `timeStamp` (in milliseconds), that you can use to distinguish them – `SyntheticEvent` correctly copies this attribute as well.

I am however not sure it is needed here.